### PR TITLE
Fix for NegativeArraySizeException in RFC5649WrapEngine

### PR DESCRIPTION
**Pull request by simon-greatrix**

### CODE REVIEW

File: core/src/main/java/org/bouncycastle/crypto/engines/RFC5649WrapEngine.java [hunk 1]
```diff
@@ -154,7 +154,7 @@ public String getAlgorithmName()
             throw new InvalidCipherTextException("unwrap data must be a multiple of 8 bytes");
         }
 
-        if (n == 1)
+        if (n <= 1)
```
Comment (simon-greatrix) [inline core/src/main/java/org/bouncycastle/crypto/engines/RFC5649WrapEngine.java:157]: If the length was zero, then the it was processed as if it was 16 bytes, which caused a runtime exception.

File: core/src/main/java/org/bouncycastle/crypto/engines/RFC5649WrapEngine.java [hunk 2]
```diff
@@ -217,7 +217,7 @@ public String getAlgorithmName()
 
         // Check the number of padded zeros
         int expectedZeros = upperBound - mli;
-        if (expectedZeros >= paddedPlaintext.length)
+        if (expectedZeros >= paddedPlaintext.length || expectedZeros<0)
```
Comment (simon-greatrix) [inline core/src/main/java/org/bouncycastle/crypto/engines/RFC5649WrapEngine.java:220]: Added catch for the expected number of zeros being negative.

File: core/src/test/java/org/bouncycastle/crypto/test/AESWrapPadTest.java [hunk 3]
```diff
@@ -162,6 +163,58 @@ public void performTest()
             rnd.nextBytes(keyToWrap);
             wrapAndUnwrap(kek, keyToWrap);
         }
+
```
Comment (simon-greatrix) [inline core/src/test/java/org/bouncycastle/crypto/test/AESWrapPadTest.java:166]: All the existing tests checked correct behaviour. I've added tests for incorrect behaviour.

File: core/src/main/java/org/bouncycastle/crypto/engines/RFC5649WrapEngine.java [hunk 4]
```diff
@@ -215,12 +215,13 @@ public String getAlgorithmName()
...
-        // Check the number of padded zeros
+        // Check the number of padding zeros
         int expectedZeros = upperBound - mli;
-        if (expectedZeros >= paddedPlaintext.length)
+        if (expectedZeros >= 8 || expectedZeros < 0)
         {
+            // We have to pick a "typical" amount of padding to avoid timing attacks.
             isValid = false;
-            expectedZeros = paddedPlaintext.length;
+            expectedZeros = 4;
```
Comment (simon-greatrix) [inline core/src/main/java/org/bouncycastle/crypto/engines/RFC5649WrapEngine.java:224]: I believe this was a timing attack. The expected number of padding zeros was from 0 to 7. By performing a verification of the entire plain text we were doing a lot more work on the comparison than is typical.

**Comment by dghgit** [comment]
Thanks for the patch! Now merged.

**Pull request closed without merging**